scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +741,336 @@ class ProcessMonitor(ScalyrMonitor):
-            self.__set_process(self.__select_process())
+        # using the historical values, calculate the aggregate
+        # there are two kinds of metrics:
+        # a) cumulative metrics - only the delta of the last 2 recorded values is used (eg cpu cycles)
+        # b) absolute metrics - the last absolute value is used
+
+        running_pids_set = set(running_pids)
+
+        for pid, process_metrics in self.__metrics_history.items():
+            for _metric, _metric_values in process_metrics.items():
+                if not self.__aggregated_metrics.get(_metric):
+                    self.__aggregated_metrics[_metric] = 0
+                if _metric.is_cumulative:
+                    if pid in running_pids_set:
+                        if len(_metric_values) > 1:
+                            # only report the cumulative metrics for more than one sample
+                            self.__aggregated_metrics[_metric] += (_metric_values[-1] - _metric_values[-2])
+                        else:
+                            self.__aggregated_metrics[_metric] = 0
+                else:
+                    if pid in running_pids_set:
+                        # absolute metric - accumulate the last reported value
+                        self.__aggregated_metrics[_metric] += _metric_values[-1]
+
+    def _remove_dead_processes(self, running_pids):
+        # once the _calculate_aggregated_metrics has calculated the running total of the
+        # metrics for the current epoch, remove the entries for the process id in the history
+        # NOTE: the removal of the contributions (if applicable) should already be done so
+        # removing the entry from the history is safe.
+
+        all_pids = self.__metrics_history.keys()
+        for _pid_to_remove in list(set(all_pids) - set(running_pids)):
+            # for all the absolute metrics, decrease the count that the dead processes accounted for
+            del self.__metrics_history[_pid_to_remove]
+
+        # if no processes are running, there is no reason to report the running metric data
+        if not running_pids:
+            self.__aggregated_metrics = {}
+
+    def get_pids_from_ps(self, subprocesses=False, ppid=None):
+        """
+        Get the list of pids running, or get the list of child processes given a parent process id
+        :param subprocesses: Get subprocesses?
+        :param ppid: parent process id, only if subprocesses is True
+        :type subprocesses: bool
+        :type ppid: int
+        :return: list of process ids
+        """
 
-        for gather in self.__gathers:
-            gather.run_single_cycle()
+        cmd = ['ps', 'ax', '-o', 'pid,command']
+        if subprocesses:
+            cmd = ["ps", "--ppid", str(ppid), "-o", "pid,command"]
+        sub_proc = Popen(cmd, shell=False, stdout=PIPE)
+        lines = sub_proc.stdout.readlines()
+        all_pids = []
+        try:
+            for line in lines:
+                line = line.strip()
```

Comment:
I know you didn't write this code, but when you update it, I believe you can greatly simplify it with a regular expression and matching groups.  Match just on two integers followed by a string that is the commandline.